contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -2423,3 +2426,59 @@ pg_extension_config_remove(PG_FUNCTION_ARGS)
 
 	PG_RETURN_VOID();
 }
+
+/*
+ * The EOMONTH function is a Transact-SQL function in SQL Server that returns 
+ * the last day of the month of a specified date, with an optional offset.
+ */
+Datum
+EOMONTH(PG_FUNCTION_ARGS)
+{
+    int year, month, day;
+    DateADT date = PG_GETARG_DATEADT(0);
+    int offset = PG_GETARG_INT32(1);
+
+    /* Convert the date to year, month, day */
+    j2date(date + POSTGRES_EPOCH_JDATE, &year, &month, &day);
+
+    /* Adjust the month based on the offset */
+    month += offset;
+    // Check if the new month is greater than 0.
+    if(month > 0)
+    {
+        /* 
+         * If it is, calculate the new year by adding the number of full years in the offset to the current year.
+         * The number of full years in the offset is (month - 1) / 12. 
+         */
+        year += (month - 1) / 12;
```

Comment:
Could you please help explain these for calculations? 
if (month > 0 )
{
year += (month - 1) / 12;
 month = (month - 1) % 12 + 1; 
 }
 else 
{
 year += month / 12 - 1;
month = month % 12 + 12;
}

Reply:
I feel several of the comments here are just restating the code. They could potentially be more helpful if they instead explained why we have the if/else here (i.e. different case with positive / minus offset). And why we need to adjust the month by one only in the positive case.

Reply:
In the expression (month - 1) / 12, the month - 1 operation is performed to account for the fact that in calendar, months are 1-indexed, meaning they start at 1 (January) and end at 12 (December).

When we add an offset to the current month, and that value is greater than 12, it means we crossed a new year. The (month - 1) / 12 operation is basically asking, "how many full cycles of 12 months are there in our offset?". So we have to think in terms of zero-based indices, hence the month - 1.

Example: If the current month is January (1) and we add an offset of 11, the result is 12 (December). In terms of full years passed, the answer is zero, even though the number is 12, because we’re still within the first year. By subtracting 1 before the division, we correctly calculate that no full year has passed.

The same logic applies to (month - 1) % 12 + 1. We subtract 1 before taking the modulus so that we're working with a zero (where January is 0 and December is 11), perform the modulus operation to find the remainder months that didn't form a full year, and then add 1 back to convert back to a 1.

Reply:
In the expression (month - 1) / 12, the month - 1 operation is performed to account for the fact that in calendar, months are 1-indexed, meaning they start at 1 (January) and end at 12 (December).

When we add an offset to the current month, and that value is greater than 12, it means we crossed a new year. The (month - 1) / 12 operation is basically asking, "how many full cycles of 12 months are there in our offset?". So we have to think in terms of zero-based indices, hence the month - 1.

Example: If the current month is January (1) and we add an offset of 11, the result is 12 (December). In terms of full years passed, the answer is zero, even though the number is 12, because we’re still within the first year. By subtracting 1 before the division, we correctly calculate that no full year has passed.

The same logic applies to (month - 1) % 12 + 1. We subtract 1 before taking the modulus so that we're working with a zero (where January is 0 and December is 11), perform the modulus operation to find the remainder months that didn't form a full year, and then add 1 back to convert back to a 1.

Reply:
Yes,I have added comments in this new revision and have explain in that.